File: LL-Cycle/ll_cycle.cpp
bool has_cycle(Node* head) {

    if (head == NULL){
        return 0;
    }

    Node *ptr1 = head;
    Node *ptr2 = head;

    while(ptr1 != NULL && ptr2 != NULL && ptr1->next){

        ptr1 = ptr1->next->next;
        ptr2 = ptr2->next;

        if(ptr1==ptr2){
            return 1;
        }

    }

    return 0;
}



/*
bool has_cycle(Node* head) {
    if (head == NULL){
        return 0;
    }
    Node* node1 = head;
    Node* node2 = head;
    node2 = node2->next;
    while((node1->data != node2->data) && (node1 != NULL) && (node2 != NULL)){
        node1 = node1->next;
        node2 = node2->next->next;
    }
    if(node1 == NULL || node2 == NULL){
        return 0;
    }

    return 1;
}
/*

File: LL-Cycle/ll_cycle.py
"""
Detect a cycle in a linked list. Note that the head pointer may be 'None' if the list is empty.


A Node is defined as: 
""" 
class Node(object):
    def __init__(self, data = None, next_node = None):
        self.data = data
        self.next = next_node


def has_cycle(head):
    node1=head
    node2=head.next
    
    while(node1 != node2 and node1 != None and node2 != None):
        node1 = head.next
        node2 = head.next.next.next
        
    if(node1 == None or node2 == None):
        return 0
    
    elif(node1 == node2):
        return 1
            
    

